pkg/admission/machines.go

```diff
@@ -0,0 +1,97 @@
+package admission
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/golang/glog"
+
+	clusterv1alpha1conversions "github.com/kubermatic/machine-controller/pkg/apis/cluster/v1alpha1/conversions"
+	"github.com/kubermatic/machine-controller/pkg/cloudprovider"
+	"github.com/kubermatic/machine-controller/pkg/providerconfig"
+	"github.com/kubermatic/machine-controller/pkg/userdata"
+
+	admissionv1beta1 "k8s.io/api/admission/v1beta1"
+	apiequality "k8s.io/apimachinery/pkg/api/equality"
+
+	clusterv1alpha1 "sigs.k8s.io/cluster-api/pkg/apis/cluster/v1alpha1"
+)
+
+const latestKubernetesVersion = "1.12.1"
+
+func (ad *admissionData) mutateMachines(ar admissionv1beta1.AdmissionReview) (*admissionv1beta1.AdmissionResponse, error) {
+
+	machine := clusterv1alpha1.Machine{}
+	if err := json.Unmarshal(ar.Request.Object.Raw, &machine); err != nil {
+		return nil, fmt.Errorf("failed to unmarshal: %v", err)
+	}
+	machineOriginal := machine.DeepCopy()
+	glog.V(4).Infof("Defaulting and validating machine %s/%s", machine.Namespace, machine.Name)
+
+	// Mutating .Spec is never allowed
+	// Only hidden exception: the machine-controller may set the .Spec.Name to .Metadata.Name
+	// because otherwise it can never add the delete finalizer as it internally defaults the Name
+	// as well, since on the CREATE request for machines, there is only Metadata.GenerateName set
+	// so we can't default it initially
+	if ar.Request.Operation == admissionv1beta1.Update {
+		oldMachine := clusterv1alpha1.Machine{}
+		if err := json.Unmarshal(ar.Request.OldObject.Raw, &oldMachine); err != nil {
+			return nil, fmt.Errorf("failed to unmarshal OldObject: %v", err)
+		}
+		if oldMachine.Spec.Name != machine.Spec.Name && machine.Spec.Name == machine.Name {
+			oldMachine.Spec.Name = machine.Spec.Name
+		}
+		if equal := apiequality.Semantic.DeepEqual(machine.Spec, oldMachine.Spec); !equal {
+			return nil, fmt.Errorf("machine.spec is immutable")
+		}
+	}
+
+	// Add type revision annotation
+	if _, ok := machine.Annotations[clusterv1alpha1conversions.TypeRevisionAnnotationName]; !ok {
+		if machine.Annotations == nil {
+			machine.Annotations = map[string]string{}
+		}
+		machine.Annotations[clusterv1alpha1conversions.TypeRevisionAnnotationName] = clusterv1alpha1conversions.TypeRevisionCurrentVersion
+	}
+
+	// Default name
+	if machine.Spec.Name == "" {
+		machine.Spec.Name = machine.Name
+	}
+
+	// Default and verify .Spec on CREATE only, its expensive and not required to do it on UPDATE
```

Comment:
It could help/improve the UX on update - but to avoid unnecessary calls to the cloud provider, we could maintain a internal cache.

The cache could be: 
```go
cache := map[string]bool{}
key := string(machine.UID)+hash(resolved-providerConfig)
```

Though this is an improvement and can be done in a follow up.

Reply:
The reality is that updating a `machine` has no effect unless you manually delete the instance at the cloud provider which is why I have added a check above that will deny any `UPDATE` request except for `.Spec.Name` because we can not default it.

This is why its not required to do it on `UPDATE`, I will add that info to the comment

Reply:
Actually it is already in the comment, see the line below: `as we disallow .Spec changes anyways`